validator/client/streaming/runner.go

```diff
@@ -0,0 +1,153 @@
+			log.Fatalf("Could not determine if beacon chain started: %v", err)
+		}
+		if err := v.WaitForSync(ctx); err != nil {
+			log.Fatalf("Could not determine if beacon node synced: %v", err)
+		}
+	}
+	if err := v.WaitForActivation(ctx); err != nil {
+		log.Fatalf("Could not wait for validator activation: %v", err)
+	}
+	// We listen to a server-side stream of validator duties in the
+	// background of the validator client.
+	go func() {
+		if err := v.StreamDuties(ctx); err != nil {
+			handleAssignmentError(err, v.CurrentSlot())
+		}
+	}()
+	for {
+		ctx, span := trace.StartSpan(ctx, "validator.processSlot")
+
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, stopping validator")
+			return // Exit if context is canceled.
+		case slot := <-v.NextSlot():
+			span.AddAttributes(trace.Int64Attribute("slot", int64(slot)))
+			deadline := v.SlotDeadline(slot)
+			slotCtx, _ := context.WithDeadline(ctx, deadline)
+			// Report this validator client's rewards and penalties throughout its lifecycle.
+			log := log.WithField("slot", slot)
+			log.WithField("deadline", deadline).Debug("Set deadline for proposals and attestations")
+			if err := v.LogValidatorGainsAndLosses(slotCtx, slot); err != nil {
+				log.WithError(err).Error("Could not report validator's rewards/penalties")
+			}
+
+			if featureconfig.Get().ProtectAttester {
+				if err := v.UpdateProtections(ctx, slot); err != nil {
+					log.WithError(err).Error("Could not update validator protection")
+				}
+			}
+
+			// Start fetching domain data for the next epoch.
+			if helpers.IsEpochEnd(slot) {
+				go v.UpdateDomainDataCaches(ctx, slot+1)
+			}
+
+			var wg sync.WaitGroup
+
+			allRoles, err := v.RolesAt(ctx, slot)
+			if err != nil {
+				log.WithError(err).Error("Could not get validator roles")
+				continue
+			}
+			for id, roles := range allRoles {
+				wg.Add(len(roles))
+				for _, role := range roles {
+					go func(role validatorRole, id [48]byte) {
+						defer wg.Done()
+						switch role {
+						case roleAttester:
+							v.SubmitAttestation(slotCtx, slot, id)
+						case roleProposer:
+							v.ProposeBlock(slotCtx, slot, id)
+						case roleAggregator:
+							v.SubmitAggregateAndProof(slotCtx, slot, id)
+						case roleUnknown:
+							log.WithField("pubKey", fmt.Sprintf("%#x", bytesutil.Trunc(id[:]))).Trace("No active roles, doing nothing")
+						default:
+							log.Warnf("Unhandled role %v", role)
+						}
+					}(role, id)
+				}
+			}
+			// Wait for all processes to complete, then report span complete.
+			go func() {
+				wg.Wait()
+				v.LogAttestationsSubmitted()
+				if featureconfig.Get().ProtectAttester {
+					if err := v.SaveProtections(ctx); err != nil {
+						log.WithError(err).Error("Could not save validator protection")
+					}
+				}
+				span.End()
+			}()
+		}
+	}
+}
+
+func handleAssignmentError(err error, slot uint64) {
+	if errCode, ok := status.FromError(err); ok && errCode.Code() == codes.NotFound {
```

Comment:
`NotFound` automatically means validator is not yet assigned?

Reply:
I think so, it comes from this server-side error:

if !ok {
			return nil, status.Errorf(codes.NotFound, "Could not find validator index for public key %#x", pubkeyBytes)
		}

Reply:
The concern here is anyone can update server side with additional error `NotFound`. This will be unknowingly impacted